src/goto-programs/goto_convert_exceptions.cpp

```diff
@@ -267,39 +266,68 @@ symbol_exprt goto_convertt::exception_flag(const irep_idt &mode)
   return symbol_exprt(id, bool_typet());
 }
 
-void goto_convertt::unwind_destructor_stack(
+/// Unwinds the destructor stack and creates destructors for each node between
+/// destructor_start_point and destructor_end_point (including the start,
+/// excluding the end).
+///
+/// If destructor_end_point isn't passed, it will unwind the whole stack.
+/// If destructor_start_point isn't passed, it will unwind from the current
+/// node.
+bool goto_convertt::unwind_destructor_stack(
   const source_locationt &source_location,
-  std::size_t final_stack_size,
   goto_programt &dest,
-  const irep_idt &mode)
+  const irep_idt &mode,
+  optionalt<node_indext> destructor_end_point,
+  optionalt<node_indext> destructor_start_point)
 {
-  unwind_destructor_stack(
-    source_location, final_stack_size, dest, targets.destructor_stack, mode);
+  return unwind_destructor_stack(
+    source_location,
+    dest,
+    targets.destructor_stack,
+    mode,
+    destructor_end_point,
+    destructor_start_point);
 }
 
-void goto_convertt::unwind_destructor_stack(
+bool goto_convertt::unwind_destructor_stack(
   const source_locationt &source_location,
-  std::size_t final_stack_size,
   goto_programt &dest,
-  destructor_stackt &destructor_stack,
-  const irep_idt &mode)
+  destructor_treet &destructor_stack,
+  const irep_idt &mode,
+  optionalt<node_indext> destructor_end_point,
+  optionalt<node_indext> destructor_start_point)
 {
-  // There might be exceptions happening in the exception
-  // handler. We thus pop off the stack, and then later
-  // one restore the original stack.
-  destructor_stackt old_stack=destructor_stack;
+  // We keep track of currently-destroyed nodes so any unwinding done
+  // by the convert() call dosen't re-destroy ones that have been done
+  // just before.
+
+  // This will only be set by the root unwind, while every other step down
+  // the call stack will add to it.
+  bool root = false;
+  if(!targets.destroyed_nodes)
```

Comment:
Let's not invent a new way of doing this, just replicate the old algorithm -- when it pops, do a "pop" ourselves by setting the current node to current->in[0], then at the end set the current node to the old curent.